lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -955,81 +993,90 @@ def determine_agent_level_common_labels
       labels
     end
 
-    # Determine the group level monitored resource and common labels shared by a
-    # collection of entries.
-    def determine_group_level_monitored_resource_and_labels(tag)
-      # Determine group level monitored resource type. For certain types,
-      # extract useful info from the tag and store those in
-      # matched_regex_group.
-      group_resource_type, matched_regex_group =
-        determine_group_level_monitored_resource_type(tag)
-
-      # Determine group level monitored resource labels and common labels.
-      group_resource_labels, group_common_labels =
-        determine_group_level_labels(group_resource_type, matched_regex_group)
-
-      group_resource = Google::Apis::LoggingV2beta1::MonitoredResource.new(
-        type: group_resource_type,
-        labels: group_resource_labels.to_h
-      )
-
-      # Freeze the per-request state. Any further changes must be made on a
-      # per-entry basis.
-      group_resource.freeze
-      group_resource.labels.freeze
-      group_common_labels.freeze
-
-      [group_resource, group_common_labels]
+    # Group the log entries by tags.
+    def group_log_entries_by_tag(chunk)
+      groups = {}
+      chunk.msgpack_each do |tag, *arr|
+        sanitized_tag = sanitize_tag(tag)
+        if sanitized_tag.nil?
+          @log.warn "Dropping log entries with invalid tag: '#{tag}'. " \
+                    'A tag should be a string with utf8 characters.'
+          next
+        end
+        groups[sanitized_tag] ||= []
+        groups[sanitized_tag].push(arr)
+      end
+      groups
     end
 
-    # Determine group level monitored resource type shared by a collection of
-    # entries.
-    # Return the resource type and tag regexp matched groups. The matched groups
-    # only apply to some resource types. Return nil if not applicable or if
-    # there is no match.
-    def determine_group_level_monitored_resource_type(tag)
-      @tag_regexp_list.each do |derived_type, tag_regexp|
-        matched_regex_group = tag_regexp.match(tag)
-        return [derived_type, matched_regex_group] if
-          matched_regex_group
+    # Once log entries are grouped by tags, we want to further group them by
+    # locally_unique_id, which is specified in the payload under the key
+    # "logging.googleapis.com/locally_unique_id". This way we can avoid
+    # redundant calls to Metadata Agent to retrieve the monitored resource
+    # corresponding to the locally_unique_id.
+    def group_log_entries_by_unique_id(entries)
+      groups = {}
+      # Log entries with no locally unique id go into a group with key
+      # NO_UNIQUE_ID_PRESENT.
+      groups[NO_UNIQUE_ID_PRESENT] = []
+      entries.each do |time, record|
+        if record.is_a?(Hash) && record.key?(DEFAULT_UNIQUE_ID_KEY)
```

Comment:
It *should* always be a `Hash`. User could mess with it though. For example they can send non-hash values via `in_forward` plugin. We used to have this check. So I kept it. We should add some logs when it's not a hash indicating we are dropping invalid log entries though.

Reply:
There is a test `test_malformed_log` that makes sure we throw away non-hash logs instead of letting the agent crash. For logs coming from `in_forward` plugin, the `record` can be anything.